compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+  }
+
+  static const char* kindToString(TupleDecl::IntentOrKind kind) {
+    switch(kind) {
+      case TupleDecl::IntentOrKind::DEFAULT_INTENT: assert(false);
+      case TupleDecl::IntentOrKind::CONST_INTENT: return "const";
+      case TupleDecl::IntentOrKind::VAR: return "var";
+      case TupleDecl::IntentOrKind::CONST_VAR: return "const var";
+      case TupleDecl::IntentOrKind::CONST_REF: return "const ref";
+      case TupleDecl::IntentOrKind::REF: return "ref";
+      case TupleDecl::IntentOrKind::IN: return "in";
+      case TupleDecl::IntentOrKind::CONST_IN: return "const in";
+      case TupleDecl::IntentOrKind::OUT: return "out";
+      case TupleDecl::IntentOrKind::INOUT: return "inout";
+      case TupleDecl::IntentOrKind::INDEX: assert(false);
+      case TupleDecl::IntentOrKind::PARAM: return "param";
+      case TupleDecl::IntentOrKind::TYPE: return "type";
+    }
+    assert(false);
+  }
+
+  static const char* kindToString(Variable::Kind kind) {
+    switch (kind) {
+      case Variable::Kind::VAR: return "var";
+      case Variable::Kind::CONST: return "const";
+      case Variable::Kind::PARAM: return "param";
+      case Variable::Kind::TYPE: return "type";
+      case Variable::Kind::REF: return "ref";
+      case Variable::Kind::CONST_REF: return "const ref";
+      case Variable::Kind::INDEX: assert(false);
+    }
+    assert(false);
+    return "";
+  }
+
+  static const char* kindToString(VisibilityClause::LimitationKind kind) {
+    switch (kind) {
+      case VisibilityClause::LimitationKind::NONE: assert(false);
+      case VisibilityClause::LimitationKind::ONLY: return "only";
+      case VisibilityClause::LimitationKind::EXCEPT: return "except";
+      case VisibilityClause::LimitationKind::BRACES: assert(false);
+      default: return "";
+    }
+  }
+
+
+  /** visit each elt of begin..end, outputting `separator` between each.
+   * `surroundBegin` and `surroundEnd` are output before and after respectively
+   * if not null */
+  template<typename It>
+  void interpose(std::ostream& os, It begin, It end, const char* separator, const char* surroundBegin=nullptr, const char* surroundEnd=nullptr) {
+    bool first = true;
+    if (surroundBegin) os << surroundBegin;
+    for (auto it = begin; it != end; it++) {
+      if (!first) os << separator;
+      printAst(os,*it);
+      first = false;
+    }
+    if (surroundEnd) os << surroundEnd;
+  }
+
+  template<typename T>
+  void interpose(std::ostream& os, T xs, const char* separator, const char* surroundBegin=nullptr, const char* surroundEnd=nullptr) {
+    interpose(os, xs.begin(), xs.end(), separator, surroundBegin, surroundEnd);
+  }
+
+
+  /*
+   * helper for printing descendants of simpleBlockLike to handle when to print
+   * the optional opening keyword and if braces should follow that keyword
+   */
+  template<typename It>
+  void printBlockWithStyle(std::ostream& os, BlockStyle style,  It begin, It end, const char* implicitOpeningKeyword=nullptr) {
```

Comment:
If we make these methods on the visitor struct then we can avoid having to pass in the `ostream` (and later the indentation level, if that's the route we choose to go).